windows/src/engine/keymanx64/keymanx64.cpp

```diff
@@ -353,18 +372,27 @@ void Shutdown()
   Keyman_StartExit();  // I3092
 
   /* Tell all threads that it is time to exit.  This is important to do before we shutdown
-    because we have got a per-thread keyboard hook that needs to be detached before we
-    lose our message hooks. */
+     because we want to try and detach from as many processes as possible so we don't
+     remain locked in memory. */
 
   SendMessageTimeout(HWND_BROADCAST, wm_keyman, KM_EXIT, 0, SMTO_NORMAL, 1000, &dwResult);  // I3092
 
   Keyman_Exit();
 
   /* Unregister those windows that we registered earlier - don't fail on error though */
+
   if(hwndController != NULL) Keyman_UnregisterControllerWindow(hwndController);   // I3758
   if(hwndControllerOwner != NULL) Keyman_UnregisterControllerWindow(hwndControllerOwner);   // I3758
 
   PostQuitMessage(0);
+
+  /* Cleanup */
+
+  SetEvent(hWatcherThreadTerminateEvent);
```

Comment:
Does cleanup work when either `hWatcherThreadTerminateEvent` or `hWatcherThread` are NULL?
(for when `CreateWatcherThread()` fails)

Reply:
Good question. Turns out `WM_DESTROY` is called even if `WM_CREATE` returns `FALSE`, so it is important to cleanup correctly. Done now!